<img style="float: left;;" src='Figures/alinco.png' /></a>

# <center> <font color= #000047> Manipulación de Datos con la librería Pandas



<img style="float: right; margin: 0px 0px 15px 15px;" src="https://numfocus.org/wp-content/uploads/2016/07/pandas-logo-300.png" width="400px" height="400px" />

> Ya vamos agregando más librerías a nuestro arsenal. En estas clases hemos revisado aspectos muy básicos de estas librerías ([NumPy](https://www.numpy.org/) y [matplotlib](https://matplotlib.org/)). En estos enlaces pueden encontrar la documentación oficial para revisar qué más pueden hacer con estas librerías. También el `help` puede ser un buen punto de partida para conocer qué más cosas pueden realizar.


Referencias:
- https://pandas.pydata.org/
- https://towardsdatascience.com/data-science-with-python-intro-to-loading-and-subsetting-data-with-pandas-9f26895ddd7f
___

In [ ]:
# Importar pandas


Los **pd.DataFrames** son los objetos por excelencia de pandas para manipular datos. Son eficientes y rápidos. Son la estructura de datos donde pandas carga los diferentes formatos de datos: cuando nuestros datos están limpios y estructurados, cada fila representa una observación, y cada columna una variable o característica. Tanto las filas como las columnas pueden tener etiquetas.

En esta clase vamos a trabajar con datos pertenecientes a ejemplos de la librería featuretools [en esta página](https://github.com/FeatureLabs/featuretools). Sin embargo para entender cómo es que se cargan estos datos, estos están disponibles en la carpeta data en formato CSV.

En esta y la siguiente clase, además de importar datos, aprenderemos a:
- seleccionar subconjuntos de datos;
- filtrar variables por categorías;
- relacionar tablas con datos complementarios;
- entre otros.

Comenzamos:

___
# 1. Importando datos

En la carpeta "data" tenemos los archivos "customers_data.csv", "sessions_data.csv", "transactions_data.csv" y "products_data.csv", cada uno de ellos conteniendo información relevante acerca de transacciones de cierta empresa.

El primer paso para comenzar a trabajar con datos es importarlos. Lo podemos hacer con la función `pd.read_csv()`:

In [5]:
# Ayuda en la función pd.read_csv()
import pandas as pd
customers = pd.read_csv('Data/customers_data.csv')
customers.head(5)

Unnamed: 0  customer_id  zip_code            join_date date_of_birth
0           0            1     60091  2011-04-17 10:48:33    1994-07-18
1           1            2     13244  2012-04-15 23:31:04    1986-08-18
2           2            3     13244  2011-08-13 15:42:34    2003-11-21
3           3            4     60091  2011-04-08 20:08:14    2006-08-15
4           4            5     60091  2010-07-17 05:27:50    1984-07-28

In [6]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Unnamed: 0     5 non-null      int64 
 1   customer_id    5 non-null      int64 
 2   zip_code       5 non-null      int64 
 3   join_date      5 non-null      object
 4   date_of_birth  5 non-null      object
dtypes: int64(3), object(2)
memory usage: 332.0+ bytes


Muy bien, concentrémonos en los datos de cliente (comenzaremos con estos datos pequeños para ilustrar las funcionalidades, luego iremos añadiendo tamaño):

In [9]:
# Importar customers_data.csv
customers = pd.read_csv('Data/customers_data.csv')
customers

Unnamed: 0  customer_id  zip_code            join_date date_of_birth
0           0            1     60091  2011-04-17 10:48:33    1994-07-18
1           1            2     13244  2012-04-15 23:31:04    1986-08-18
2           2            3     13244  2011-08-13 15:42:34    2003-11-21
3           3            4     60091  2011-04-08 20:08:14    2006-08-15
4           4            5     60091  2010-07-17 05:27:50    1984-07-28

Bien, ya tenemos los datos cargados, sin embargo, se ve algo raro. 

¿Qué es esta columna "Unnamed: 0"? (abrir el CSV)

Para especificar que esta columna corresponde al índice, podemos usar el argumento `index_col`.

In [11]:
# Importar customers_data.csv, haciendo uso del argumento index_col
customers = pd.read_csv('Data/customers_data.csv', index_col=0)
customers

customer_id  zip_code            join_date date_of_birth
0            1     60091  2011-04-17 10:48:33    1994-07-18
1            2     13244  2012-04-15 23:31:04    1986-08-18
2            3     13244  2011-08-13 15:42:34    2003-11-21
3            4     60091  2011-04-08 20:08:14    2006-08-15
4            5     60091  2010-07-17 05:27:50    1984-07-28

In [8]:
# Tipo de lo que importamos
type(customers)

pandas.core.frame.DataFrame

Ahora nuestro DataFrame luce bien. Sin embargo, podríamos querer indizar nuestro DataFrame directamente por el id de cliente, utilizando el método `set_index()`:

In [ ]:
# Usar el método set_index para indizar por el id de cliente
customers.set_index('customer_id')
customers

zip_code            join_date date_of_birth
customer_id                                             
1               60091  2011-04-17 10:48:33    1994-07-18
2               13244  2012-04-15 23:31:04    1986-08-18
3               13244  2011-08-13 15:42:34    2003-11-21
4               60091  2011-04-08 20:08:14    2006-08-15
5               60091  2010-07-17 05:27:50    1984-07-28

Podemos ver que pandas no convierte **customer_id** en el índice permanente del DataFrame, a menos que se lo especifiquemos. Para especificarlo, debemos usar el argumento `inplace`

In [15]:
# Usar el método set_index para indizar por el id de cliente, usando el argumento inplace
# customers = customers.set_index("customer_id")

#customers.set_index('customer_id', inplace=True)
#customers

___
# 2. Indización y selección de datos

Hay muchas formas de las cuales podemos seleccionar datos de DataFrames. Veremos, de acuerdo al artículo al final de este documento, la forma basada en corchetes ([]) y en los métodos `loc()` y `iloc()`.

Con los corchetes, podemos seleccionar ciertas filas, o bien, ciertas columnas. 

Para una selección de filas, podemos usar el indizado como en las listas: [start_index:end_index:step], recordando que el `end_index` no es inclusivo.

Por ejemplo, seleccionar los clientes en las primeras dos filas:

In [16]:
customers[:2]

zip_code            join_date date_of_birth
customer_id                                             
1               60091  2011-04-17 10:48:33    1994-07-18
2               13244  2012-04-15 23:31:04    1986-08-18

Ahora, seleccionar del tercer cliente en adelante:

In [17]:
customers[3:]

zip_code            join_date date_of_birth
customer_id                                             
4               60091  2011-04-08 20:08:14    2006-08-15
5               60091  2010-07-17 05:27:50    1984-07-28

Finalmente, seleccionar los clientes en las filas impares:

In [18]:
customers[::2]

zip_code            join_date date_of_birth
customer_id                                             
1               60091  2011-04-17 10:48:33    1994-07-18
3               13244  2011-08-13 15:42:34    2003-11-21
5               60091  2010-07-17 05:27:50    1984-07-28

Similarmente, para una selección de columnas, podemos usar una lista con los nombres de las columnas requeridas. Si pasamos solo una columna como un string en lugar de una lista, el resultado será un objeto **pd.Series**:

In [19]:
# Seleccionar la columna date_of_birth como Series
customers['date_of_birth']

customer_id
1    1994-07-18
2    1986-08-18
3    2003-11-21
4    2006-08-15
5    1984-07-28
Name: date_of_birth, dtype: object

In [20]:
# Tipo
type(customers['date_of_birth'])

pandas.core.series.Series

Las Series de pandas son arreglos unidimensionales que pueden ser etiquetados. En algunos casos, querremos seleccionar solo una columna pero mantener esta selección en un DataFrame. En este caso podemos pasar una lista con un solo elemento:

In [21]:
# Seleccionar la columna date_of_birth como DataFrame
customers[['date_of_birth']]

date_of_birth
customer_id              
1              1994-07-18
2              1986-08-18
3              2003-11-21
4              2006-08-15
5              1984-07-28

In [22]:
# Tipo
type(customers[['date_of_birth']])

pandas.core.frame.DataFrame

Finalmente, seleccionamos dos columnas:

In [ ]:
# Seleccionar las columnas date_of_birth y zip_code
customers.loc[:,['date_of_birth','zip_code']]


date_of_birth  zip_code
customer_id                        
1              1994-07-18     60091
2              1986-08-18     13244
3              2003-11-21     13244
4              2006-08-15     60091
5              1984-07-28     60091

In [30]:
customers.iloc[:,[2,0]]

date_of_birth  zip_code
customer_id                        
1              1994-07-18     60091
2              1986-08-18     13244
3              2003-11-21     13244
4              2006-08-15     60091
5              1984-07-28     60091

Muy bien, ya vimos que los corchetes son útiles. También existen los poderosos métodos `loc` y `iloc`, que nos dan el poder de seleccionar ambos a la vez: columnas y filas.

¿En qué se diferencian?

- El método `loc` nos permite seleccionar filas y columnas de nuestros datos basados en etoquetas. Primero, se deben especificar las etiquetas de las filas, y luego las de las columnas.

- El método `lioc` nos permite hacer lo mismo pero basado en índices enteros de nuestro DataFrame (como si fueran matrices).

Como antes, si queremos seleccionar todas las filas, o columnas, simplemente escribimos `:` en el lugar adecuado.

Mejor con ejemplos:

Para un mejor entendimiento de esta parte, reetiquetaremos la numeración de clientes con una "numeración" alfabética. Es decir: 1-A, 2-B, ..., 5-E.

In [32]:
# Resetear índice en el lugar
customers.reset_index(inplace=True)
customers

index  customer_id  zip_code            join_date date_of_birth
0      0            1     60091  2011-04-17 10:48:33    1994-07-18
1      1            2     13244  2012-04-15 23:31:04    1986-08-18
2      2            3     13244  2011-08-13 15:42:34    2003-11-21
3      3            4     60091  2011-04-08 20:08:14    2006-08-15
4      4            5     60091  2010-07-17 05:27:50    1984-07-28

Ahora sí.

Seleccionemos el primer cliente con ambos métodos:

In [39]:
# Primer cliente con loc
customers.loc['0']

KeyError: '0'

In [37]:
# Primer cliente con iloc
customers.iloc[0:]

index  customer_id  zip_code            join_date date_of_birth
0      0            1     60091  2011-04-17 10:48:33    1994-07-18
1      1            2     13244  2012-04-15 23:31:04    1986-08-18
2      2            3     13244  2011-08-13 15:42:34    2003-11-21
3      3            4     60091  2011-04-08 20:08:14    2006-08-15
4      4            5     60091  2010-07-17 05:27:50    1984-07-28

Nos dan series, ¿y si queremos DataFrames?

In [ ]:
# Primer cliente con loc en DataFrame


In [ ]:
# Primer cliente con iloc en DataFrame


Ahora, seleccionemos los clientes A y C con ambos métodos:

In [ ]:
# Clientes A y C con loc


In [ ]:
# Clientes A y C con iloc


Ahora, de los clientes B y E, queremos su zip_code y su join_date:

In [ ]:
# loc


In [ ]:
# iloc


Ahora, queremos solo zip_code y join_date, pero de todos los clientes:

In [ ]:
# loc


In [ ]:
# iloc


¿Qué tal? Ya tenemos varias formas de seleccionar e indexar ciertos datos.

Esto es, sin duda, muy útil. Por otra parte, muchas veces queremos obtener cierta información (clientes, en nuestro ejemplo) que cumplan algunos requisitos. Por ejemplo:
- que sean mayores de 18 años, o
- que su antiguedad en la plataforma sea menor a seis meses, o
- que residan en cierta zona,
- entre otros.

Para ello utilizamos los operadores de comparación (==, >, <, >=, <=, !=).

___
# 3. Filtrado de datos

Los operadores de comparación pueden ser utilizados con Series de pandas. 

Esto resulta ser súper útil para filtrar datos con ciertas condiciones específicas (esto lo veremos enseguida). 

Por ahora, veremos que al usar operadores de comparación con Series, el resultado será Series booleanas. Es decir, cada elemento de dichas series será `True` si la condición se satisface, y `Falso` de lo contrario.

Una vez tenemos la serie boolena, podemos aplicarla para la selección de filas, obteniendo un DataFrame filtrado como resultado.

**Ejemplos**

¿Cuáles clientes son mayores de 18 años?

In [40]:
customers

index  customer_id  zip_code            join_date date_of_birth
0      0            1     60091  2011-04-17 10:48:33    1994-07-18
1      1            2     13244  2012-04-15 23:31:04    1986-08-18
2      2            3     13244  2011-08-13 15:42:34    2003-11-21
3      3            4     60091  2011-04-08 20:08:14    2006-08-15
4      4            5     60091  2010-07-17 05:27:50    1984-07-28

In [41]:
# Adult clients
customers.iloc[0,-1]

'1994-07-18'

¿Cuáles clientes se registraron después del primero de Junio del 2011?

In [ ]:
# Recent joins


¿Cuáles clientes viven en la zona con CP 13244?

In [ ]:
# Zone


Entonces ya tenemos las series booleanas de las que hablábamos. Ahora, ¿cómo podemos usar esto para obtener la información de los clientes que cumplen estas características?

In [ ]:
# Filtrar por clientes adultos


In [ ]:
# Filtrar por registros recientes


In [ ]:
# Filtrar por zona


**¡Excelente!**

Los filtrados que acabamos de hacer fueron de una sola condición. Ahora, ¿cómo hacer filtrados con múltiples condiciones?

## 3.1 Filtrado con múltiples condiciones

Cuando queremos filtrar datos con múltiples condiciones, podemos usar operadores booleanos (`and`, `or`, `not`), solo con una pequeña modificación. Usamos:

- `&` en lugar de `and`,
- `|` en lugar de `or`, y
- `~` en lugar de `not`.

**Ejemplos**

Obtener la información de los clientes adultos que vivan en la zona con código postal 13244

Obtener la información de los clientes que sean adultos o que se hayan registrado luego del 6 de Junio del 2011

Obtener la información de los clientes que **NO** se hayan registrado en 2011

De esta manera, podemos crear condiciones tan complejas y restrictivas como queramos para filtrar nuestros datos.

Útil, ¿no?